File: server/src/main/java/com/uiptv/util/SerialAttributeParser.java
package com.uiptv.util;

import java.text.Normalizer;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class SerialAttributeParser implements StalkerAttributeParser {
    private static final Pattern SERIAL_PATTERN = Pattern.compile("(?:sn|serial).*?([A-F0-9]{10,32})\\b", Pattern.CASE_INSENSITIVE);

    @Override
    public String parse(String line) {
        String normalizedLine = Normalizer.normalize(line, Normalizer.Form.NFKD);
        Matcher matcher = SERIAL_PATTERN.matcher(normalizedLine);
        if (matcher.find()) {
            return matcher.group(1);
        }
        return null;
    }

    @Override
    public StalkerAttributeType getAttributeType() {
        return StalkerAttributeType.SERIAL;
    }
}

File: server/src/main/kotlin/com/uiptv/util/SerialAttributeParser.kt
package com.uiptv.util

import com.uiptv.api.StalkerAttributeParser
import java.text.Normalizer
import java.util.regex.Pattern

class SerialAttributeParser : StalkerAttributeParser {
    override fun parse(line: String): String? {
        val normalizedLine = Normalizer.normalize(line, Normalizer.Form.NFKD)
        val matcher = SERIAL_PATTERN.matcher(normalizedLine)
        return if (matcher.find()) matcher.group(1) else null
    }

    override fun getAttributeType(): StalkerAttributeType = StalkerAttributeType.SERIAL

    private companion object {
        val SERIAL_PATTERN: Pattern =
            Pattern.compile("(?:sn|serial).*?([A-F0-9]{10,32})\\b", Pattern.CASE_INSENSITIVE)
    }
}
